plugins/jabra-file/fu-jabra-file-firmware.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2023 GN Audio A/S
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-jabra-file-firmware.h"
+
+struct _FuJabraFileFirmware {
+	FuArchiveFirmware parent_instance;
+	guint16 dfu_pid;
+};
+
+G_DEFINE_TYPE(FuJabraFileFirmware, fu_jabra_file_firmware, FU_TYPE_FIRMWARE)
+
+static void
+fu_jabra_file_firmware_export(FuFirmware *firmware, FuFirmwareExportFlags flags, XbBuilderNode *bn)
+{
+	FuJabraFileFirmware *self = FU_JABRA_FILE_FIRMWARE(firmware);
+	fu_xmlb_builder_insert_kx(bn, "dfu_pid", self->dfu_pid);
+}
+
+static gboolean
+fu_jabra_file_firmware_parse_info(FuJabraFileFirmware *self, XbSilo *silo, GError **error)
+{
+	const gchar *version;
+	const gchar *dfu_pid_str;
+	guint64 val = 0;
+	g_autoptr(XbNode) dfu_pid = NULL;
+	g_autoptr(XbNode) build_vector = NULL;
+
+	build_vector = xb_silo_query_first(silo, "buildVector", error);
+	if (build_vector == NULL)
+		return FALSE;
+
+	version = xb_node_get_attr(build_vector, "version");
+	if (version == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_DATA,
+				    "buildVector version missing");
+		return FALSE;
+	}
+	fu_firmware_set_version(FU_FIRMWARE(self), version);
+	dfu_pid = xb_silo_query_first(silo, "buildVector/targetUsbPids", error);
+	if (dfu_pid == NULL)
+		return FALSE;
+	dfu_pid_str = xb_node_query_text(dfu_pid, "usbPid", error);
+	if (dfu_pid_str == NULL)
+		return FALSE;
+	if (!fu_strtoull(dfu_pid_str, &val, 0x0, 0xFFFF, FU_INTEGER_BASE_AUTO, error)) {
+		g_prefix_error(error, "cannot parse usbPid of %s: ", dfu_pid_str);
+		return FALSE;
+	}
+	self->dfu_pid = (guint16)val;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_firmware_parse(FuFirmware *firmware,
+			     GInputStream *stream,
+			     gsize offset,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	FuJabraFileFirmware *self = FU_JABRA_FILE_FIRMWARE(firmware);
+	g_autoptr(FuFirmware) firmware_archive = fu_archive_firmware_new();
+	g_autoptr(FuFirmware) img_xml = NULL;
+	g_autoptr(GBytes) img_blob = NULL;
+	g_autoptr(GPtrArray) files = NULL;
+	g_autoptr(XbBuilder) builder = xb_builder_new();
+	g_autoptr(XbBuilderSource) source = xb_builder_source_new();
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(FuFirmware) upgrade_archive = NULL;
+	g_autoptr(GBytes) upgrade_blob = NULL;
```

Comment:
normally i sort these

Reply:
any particular kind of sorting? I do agree it looks a bit messy

Reply:
My editor it's ctrl+1 and it sorts it by alphacase :)